dev/bots/analyze-sample-code.dart

```diff
@@ -38,174 +43,394 @@
+
+  /// A RegExp that matches a Dart constructor.
+  static final RegExp _constructorRegExp = RegExp(r'[A-Z][a-zA-Z0-9<>.]*\(');
+
+  /// The temporary directory where all output is written. This will be deleted
+  /// automatically if there are no errors.
+  Directory _tempDir;
+
+  /// The package directory for the flutter package within the flutter root dir.
+  final Directory _flutterPackage;
+
+  /// A serial number so that we can create unique expression names when we
+  /// generate them.
+  int _expressionId = 0;
+
+  /// The exit code from the analysis process.
+  int _exitCode = 0;
+
+  // Once the snippets tool has been precompiled by Dart, this contains the AOT
+  // snapshot.
+  String _snippetsSnapshotPath;
+
+  /// Finds the location of the snippets script.
+  String get _snippetsExecutable {
+    final String platformScriptPath = path.dirname(Platform.script.toFilePath());
+    return path.canonicalize(path.join(platformScriptPath, '..', 'snippets', 'lib', 'main.dart'));
+  }
+
+  static List<File> _listDartFiles(Directory directory, {bool recursive = false}) {
+    return directory.listSync(recursive: recursive, followLinks: false)
+      .whereType<File>()
+      .where((File file) => path.extension(file.path) == '.dart').toList();
+  }
+
+  /// Computes the headers needed for each sample file.
+  List<Line> get headers {
+    if (_headers == null) {
+      final List<String> buffer = <String>[];
+      buffer.add('// generated code');
+      buffer.add('import \'dart:async\';');
+      buffer.add('import \'dart:convert\';');
+      buffer.add('import \'dart:math\' as math;');
+      buffer.add('import \'dart:typed_data\';');
+      buffer.add('import \'dart:ui\' as ui;');
+      buffer.add('import \'package:flutter_test/flutter_test.dart\';');
+      for (File file in _listDartFiles(Directory(_defaultFlutterPackage))) {
+        buffer.add('');
+        buffer.add('// ${file.path}');
+        buffer.add('import \'package:flutter/${path.basename(file.path)}\';');
+      }
+      _headers = buffer.map<Line>((String code) => Line(code)).toList();
     }
-    assert(!code.any((String line) => line.contains('\n')));
-    yield* code;
-    if (postamble != null) {
-      assert(!postamble.contains('\n'));
-      yield postamble;
+    return _headers;
+  }
+
+  List<Line> _headers;
+
+  /// Checks all the samples in the Dart files in [_flutterPackage] for errors.
+  int checkSamples() {
+    _exitCode = 0;
+    Map<String, List<AnalysisError>> errors = <String, List<AnalysisError>>{};
+    try {
+      final Map<String, Section> sections = <String, Section>{};
+      final Map<String, Snippet> snippets = <String, Snippet>{};
+      _extractSamples(sections, snippets);
+      errors = _analyze(_tempDir, sections, snippets);
+    } finally {
+      if (errors.isNotEmpty) {
+        stderr.writeln('Kept temporary directory ${_tempDir.path} because there were errors.\n');
```

Comment:
Does it help to keep the directory in this case?  Once we've written the errors and line numbers out, can't we just go back to the original file?

Reply:
In particular Windows limits how many of these you can create before your process isn't allowed to create anymore

Reply:
Good point, I'll remove that part and just always nuke the temp dir.

The Windows thing wouldn't matter: It only creates one temp dir.